core/cap-0047.md

```diff
@@ -186,93 +183,67 @@ index 23918226d..3ed9625bc 100644
-+    opaque contractCode<256000>;
-+
-+    union switch (int v)
-+    {
-+    case 0:
-+        void;
-+    }
-+    ext;
-+};
-+
-+struct CreateContractTransactionEnvelope
-+{
-+    CreateContractTransaction tx;
-+
-+    /* Each decorated signature is a signature over the SHA256 hash of
-+    * a TransactionSignaturePayload */
-+    DecoratedSignature signatures<20>;
-+};
-+
- /* A TransactionEnvelope wraps a transaction with signatures. */
- union TransactionEnvelope switch (EnvelopeType type)
- {
-@@ -742,6 +789,8 @@ case ENVELOPE_TYPE_TX:
-     TransactionV1Envelope v1;
- case ENVELOPE_TYPE_TX_FEE_BUMP:
-     FeeBumpTransactionEnvelope feeBump;
-+case ENVELOPE_TYPE_TX_CREATE_CONTRACT_TX:
-+    CreateContractTransactionEnvelope create;
- };
- 
- struct TransactionSignaturePayload
-@@ -754,6 +803,8 @@ struct TransactionSignaturePayload
-         Transaction tx;
-     case ENVELOPE_TYPE_TX_FEE_BUMP:
-         FeeBumpTransaction feeBump;
-+    case ENVELOPE_TYPE_TX_CREATE_CONTRACT_TX:
-+        CreateContractTransaction create;
-     }
-     taggedTransaction;
- };
 ```
 
 ## Semantics
 
-### CreateContractTransaction
-`CreateContractTransaction` will create a new `ContractDataEntry`
+### Creating a contract using InvokeHostFunctionOp
+This CAP specifies host function to create a contract, and it can be called
+using the `InvokeHostFunctionOp` operation from
+(https://github.com/stellar/stellar-protocol/blob/master/core/cap-0052.md).
+
+`create_contract` will create a new `ContractDataEntry`
 (https://github.com/stellar/stellar-protocol/blob/master/core/cap-0053.md) with
 a `SCV_STATIC` key type, and `SCS_LEDGER_KEY_CONTRACT_CODE_WASM` key value. The
 contractID will be the SHA-256 hash of a `HashIDPreimage`, with type
-`ENVELOPE_TYPE_CONTRACT_ID`, which contains the source account that created the
-contract, and a salt provided by the user. The transaction will fail if a
-contract with the calculated contractID already exists.
+`ENVELOPE_TYPE_CONTRACT_ID_FROM_KEY`, which contains a public key, and a salt
+provided by the user. The `create_contract` host function also does some
+validation to ensure the contract and contractID are created securely. It uses
+the `check` functions and `Identifier` defined in CAP-0052 to verify the
+identity of the contract creator. 
+
+The transaction will fail if a contract with the calculated contractID already
+exists.
 
 TODO: Reserve requirement for `ContractDataEntry`?
+TODO: Another envelope type for contractIDs from accountIDs.
+
+```rust
+// Define CREATE_CONTRACT_SEPARATOR =
+//     sha256("create_contract(nonce: u256, contract: Vec<u8>, salt: u256, key: u256, creator: Identifier, sig: Vec<u8>)"),
+// Calls check with
+//     nonce = nonce
+//     params = vec![CREATE_SEPARATOR, salt, creator, sha256(contract)]
+//     key = key
+//     sig = sig
+// Creates a new contract.
+fn create_contract(nonce: u256, contract: Vec<u8>, salt: Vec<u8>, key: u256, creator: Identifier, sig: Vec<u8>) -> ContractID;
```

Comment:
What's the difference between `key` and `creator`? Same thing below. Also I don't think there is function overloading, so these should probably have different names.

Reply:
I removed creator and the other function.